content/posts/networkx/vf2pp/graph-iso-vf2pp/index.md

```diff
@@ -0,0 +1,247 @@
+
+- **Optimal node ordering**, which avoids following unfruitful branches that will result in infeasible states. We make sure that the nodes that have the biggest possibility to match are accessed first.
+- **Implementation in a non-recursive manner**, avoiding Python's maximum recursion limit while also reducing function call overhead.
+- **Caching** of both node degrees and nodes per degree in the beginning, so that we don't have to access those features in every degree check. For example, instead of doing
+
+```python
+res = []
+for node in G2.nodes():
+    if G1.degree[u] == G2.degree[node]:
+        res.append(node)
+# do stuff with res ...
+```
+
+to get the nodes of same degree as u (which happens a lot of times in the implementation), we just do:
+
+```python
+res = G2_nodes_of_degree[G1.degree[u]]
+# do stuff with res ...
+```
+
+where "G2_nodes_of_degree" stores set of nodes for a given degree. The same is done with node labels.
+
+- **Extra shrinking of the candidate set for each node** by adding more checks in the candidate selection method and removing some from the feasibility checks. In simple terms, instead of checking a lot of conditions on a larger set of candidates, we check fewer conditions but on a more targeted and significantly smaller set of candidates.
+  For example, in this code:
+
+```python
+candidates = set(G2.nodes())
+for candidate in candidates:
+    if feasibility(u, candidate):
+        do_stuff()
+```
+
+we take a huge set of candidates, which results in poor performance due to maximizing calls of "feasibility", thus performing
+the feasibility checks in a very large set. Now compare that to the following alternative:
+
+```python
+candidates = [
+    n
+    for n in G2_nodes_of_degree[G1.degree[u]].intersection(
+        G2_nodes_of_label[G1_labels[u]]
+    )
+]
+for candidate in candidates:
+    if feasibility(u, candidate):
+        do_stuff()
+```
+
+Immediately we have drastically reduced the number of checks performed and calls to the function, as now we only apply them to nodes of the same degree and label as $u$. This is a simplification for demonstration purposes. In the actual implementation there are more checks and extra shrinking of the candidate set.
+
+## Demo
+
+Let's demonstrate our **VF2++** solver on a real graph. We are going to use the graph from the Graph Isomorphism wikipedia.
+
+<p float="center">
+  <img src="G1.png" width="200" height="200">
+  <img src="G2.png" width="395" height="250">
+</p>
+
+To use VF2++, we can do the following.
+
+```python
+import networkx as nx
+
+# edges of the graph
+edges = [
+    (1, 2),
+    (1, 5),
+    (1, 4),
+    (4, 8),
+    (2, 3),
+    (2, 6),
+    (3, 4),
+    (3, 7),
+    (5, 6),
+    (5, 8),
+    (6, 7),
+    (7, 8),
+]
+
+# map the nodes between the two graphs
+mapped_nodes = {1: "a", 2: "h", 3: "d", 4: "i", 5: "g", 6: "b", 7: "j", 8: "c"}
+
+# create the two networkx graphs
+G1 = nx.Graph(edges)
+G2 = nx.relabel_nodes(G1, mapped_nodes)
```

Comment:
From a pedagogical perspective, creating the second graph via a mapping from the first sort of seems like cheating, as then it is immediately obvious the graphs are isomorphic! It's almost like you're giving the answer away at the very beginning of the example :)

Even though it's more tedious, I think this could be improved by constructing both graphs manually. This doubles as a nice opportunity to show how graphs can be constructed from edge lists in networkx:

````suggestion
Let's start by constructing the graphs from the image above. We'll call
the graph on the left `G` and the graph on the left `H`:

```python
import networkx as nx

G = nx.Graph(
    [
        ("a", "g"),
        ("a", "h"),
        ("a", "i"),
        ("g", "b"),
        ("g", "c"),
        ("b", "h"),
        ("b", "j"),
        ("h", "d"),
        ("c", "i"),
        ("c", "j"),
        ("i", "d"),
        ("d", "j"),
    ]
)

H = nx.Graph(
    [
        (1, 2),
        (1, 5),
        (1, 4),
        (2, 6),
        (2, 3),
        (3, 7),
        (3, 4),
        (4, 8),
        (5, 6),
        (5, 8),
        (6, 7),
        (7, 8),
    ]
)
```
````